dashboard/dashboard.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package dashboard
+
+//go:generate go-bindata -nometadata -o assets.go -prefix assets -pkg dashboard assets
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rcrowley/go-metrics"
+	"golang.org/x/net/websocket"
+	"html/template"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	procSampleLimit = 200
+	memSampleLimit  = 200
+)
+
+var (
+	nextId uint32 = 0 // Next connection id
+)
+
+type dashboard struct {
+	config *Config
+
+	listener net.Listener
+	index    []byte    // Index page to serve up on the web
+	conns    []*client // Currently live websocket connections
+
+	mtrcs *mtrcs  `json:",omitempty"`
+	stats *status `json:",omitempty"`
+
+	lock sync.RWMutex // Lock protecting the dashboard's internals
+}
+
+type client struct {
+	conn   *websocket.Conn // Particular live websocket connection
+	logger log.Logger      // Logger for the particular live websocket connection
+}
+
+type mtrcs struct {
+	Processor []*data `json:"proc,omitempty"`
+	Memory    []*data `json:"mem,omitempty"`
+}
+
+type data struct {
+	T     int     `json:"time,omitempty"`
+	Value float64 `json:"value,omitempty"`
+}
+
+type status struct {
+	Peers int
+	Block int
+}
+
+func NewDashboard(config *Config) (*dashboard, error) {
+	log.Trace("NewDashboard() called")
+
+	dashboard := &dashboard{
+		config: config,
+		mtrcs:  &mtrcs{},
+	}
+
+	if config.Assets == "" {
+		tmpl, err := Asset("dashboard.html")
+		if err != nil {
+			return nil, err
+		}
+
+		website := new(bytes.Buffer)
+		if err = template.Must(template.New("").Parse(string(tmpl))).Execute(website, map[string]interface{}{
+			"procSampleLimit": procSampleLimit,
+			"memSampleLimit":  memSampleLimit,
+		}); err != nil {
+			log.Crit("Failed to render the dashboard template", "err", err)
+		}
+
+		dashboard.index = website.Bytes()
+		return dashboard, nil
+	}
+
+	//TODO case: DashboardAssetsFlag is set
```

Comment:
ProTip: If you add a `TODO` comment, it's a lot more useful if you add it like `TODO (kurkomisi): bla bla bla`, because anyone later reading your code and seeing the TODO knows who to ask about this particular issue.